### DISCUSSION TITLE
OAuth Settings per collection

### DISCUSSION THREAD

**Discussion by pnmcosta** [pocketbase/pocketbase#3941] [Q&A]
Hi,

I have a new use case where I must use different Facebook APPs to login different auth collection users, e.g. Users > Facebook App 1, Members > Facebook App 2.

Because currently it is only possible to setup OAuth client/secret's globaly, I've duplicated the facebook oauth (like the OIDC's) and this works fine, I just call provider `facebook2` on my members collections. I've updated my fork with it https://github.com/pocketbase/pocketbase/compare/master...pnmcosta:pocketbase:master

However, on latest version I noticed some changes to oauth, and I wonder is there a simpler way to achieve this? I'd be ok with just injecting settings manually or something.

If you think this is a valid use case, I'm happy to contribute if you'd like.

Thanks,
Pedro

**Comment by ganigeorgiev** [comment]
I'm not sure that I understand your use case but OAuth2 configurations per collection were requested before. For now they are left aside as I wasn't sure whether it would be worth complicating the setup.

The recent OAuth2 changes allow a single OAuth2 user to authenticate in multiple auth collections, aka. changes the previous `provider,providerId` unique pair constraint to `collectionId,provider,providerId`.

**Reply by pnmcosta** [reply]
> OAuth2 configurations per collection 

This is exactly what would enable my use case.

I need to use different Facebook Apps (client/secret) for different collections. E.g. Accounts collection authenticates with my Facebook App A and Members collections with my Facebook App B.

I was thinking that we could configure the collection oauth2 override on its settings (with a override toggle to show client/secret inputs):
![image](https://github.com/pocketbase/pocketbase/assets/550684/5cd1ec86-2ab0-4eca-98cd-08367ab2358d)

And perhaps store on the app settings per `provider/collectionId`, like so:

```
    "facebook": {
        "enabled": true,
        "clientId": "12",
        "clientSecret": "abc",
        "authUrl": "",
        "tokenUrl": "",
        "userApiUrl": ""
    },
    "facebook/collection1id": {
        "enabled": true,
        "clientId": "34",
        "clientSecret": "edf",
        "authUrl": "",
        "tokenUrl": "",
        "userApiUrl": ""
    },
    "facebook/collection2id": {
        "enabled": true,
        "clientId": "56",
        "clientSecret": "gbn",
        "authUrl": "",
        "tokenUrl": "",
        "userApiUrl": ""
    }
```

but this IMO feels to much of a hack, and settings marshalling could be tricky. One thing is for sure, the SDK api is collection driven and we send the provider, so when retrieving the auth methods for the collection it could resolve the overriden values.

on the callback I'm not so sure and would need to dig deeper, but I'm sure we can pass along the collection id in state (if it doesn't already do).

Let me know your thoughts?

**Reply by ganigeorgiev** [reply]
It is certainly doable, but I don't want to dedicate time on this at the moment.

I'll consider the use case with the planned internals refactoring after https://github.com/pocketbase/pocketbase/discussions/3271 (_there are also plans to allow user to register OAuth2 providers externally without having to modify the main package which to some extend could eventually also cover your use case_).

**Reply by ganigeorgiev** [reply]
I've moved the OAuth2 configurations to be per auth collection as part of the ongoing refactoring but after I started documenting the changes I realized that I'm no longer sure that I understand the use case here and whether it was actually needed.

It also complicates a little the migrations because it is not clear whether the OAuth2 configurations (clientId and clientSecret) should be included in the collection export.

@pnmcosta Could you elaborate a little more on your use case and why do you need separate Facebook apps for each auth collection?

**Reply by pnmcosta** [reply]
Of course, Facebook has a lot of restrictions on data use, and their Products/API's you'll be using, depending on the type of user you're targeting with your Facebook app, see https://developers.facebook.com/docs/development/create-an-app/app-dashboard/app-types/ for more info.

> App types cannot be changed. If your app needs products, permissions, or features that are unavailable to its current type you must create a new app with a different type instead.

In our specific case, we cannot have only one as they would conflict with Facebook's own rules and wouldn't get approved. So we have to have 2 distinct Facebook apps, one for Users (Business) and another for Members (Consumers) and both should be able to authenticate to our product.

> I've moved the OAuth2 configurations to be per auth collection

This would be great for our specific use case, but I understand if it may not suit all. Would it be possible/easier to be able configure more than one instance of the same OAuth provider instead, and let the use of the SDK dictate which provider instance and collection you want to use at auth time?

**Reply by pnmcosta** [reply]
> it is not clear whether the OAuth2 configurations (clientId and clientSecret) should be included in the collection export

IMO, I don't think it should, at least the client secret due to it's sensibility. I think (I could be wrong) that it would be ok if collection oauth settings (like I described in my original comment) are not exported at all (AFAIK you can't export current app settings either), and you'll always have to reconfigure if you import a collection. 

It's totally up to you of course.

**Reply by ganigeorgiev** [reply]
Thank you for elaborating.

My only problem with storing them per collection is that it introduces some caveats around the collections serialization (e.g. redacting the clientId/clientSecret) that will also require extra documentation and could be confusing, but for now I'll leave it for later as a final task to experiment a little bit more with it before deciding.

> Would it be possible/easier to be able configure more than one instance of the same OAuth provider instead ...

You'll be able to do that no matter where the OAuth2 configs will be stored.

There will be support for registering custom OAuth2 providers programmatically by modifying the global `auth.Providers` map (similar to how usually DB drivers are registered), or in other words you'll be able to do something like:

```go
auth.Providers["facebook2"] = func() auth.Provider {
  // return a new blank provider...
  return auth.NewFacebookProvider()
}

// or if it is exact copy
auth.Providers["facebook2"] = auth.Providers[auth.NameFacebook]
```

**Reply by pnmcosta** [reply]
Awesome, I think your example might work for my use case.

Just one clarifying question, would I be able to set the client id/secret on the constructor as option(s), or how would you envision being able to set those for a `facebook2`?

> but for now I'll leave it for later as a final task to experiment a little bit more with it before deciding.

That's fine, I really don't need an answer now, only when you're ready.

**Reply by ganigeorgiev** [reply]
> would I be able to set the client id/secret on the constructor as option(s)

You can set the clientId/clientSecret as fixed keys with the factory function but you'll also have the option to load it dynamically with the OAuth2Config struct (_this will be either as part of the collection options or through the app settings_). Or in other words the "facebook2" provider is treated as a standalone provider like the other ones. The only thing that you'll not able to do is to specify custom UI controls for it but that will be possible with the next stage of the refactoring after the Admin UI update.

**Reply by pnmcosta** [reply]
Brilliant, I think that would suffice for my use case then. I manage my own config/settings via env vars/secrets anyway so I wouldn't require any UI for it.

Many thanks for looking at this so diligently, I can't wait for a preview of the refactoring 🙌
